contrib/containers/ci/Dockerfile

```diff
@@ -44,15 +47,71 @@ RUN set -ex; \
     libxcb-xinerama0 \
     libxcb-xkb1 \
     libxkbcommon-x11-0 \
+    linux-libc-dev \
     lsb-release \
     software-properties-common \
+    tzdata \
     unzip \
     wget \
     m4 \
     pkg-config \
     zlib1g-dev \
     && rm -rf /var/lib/apt/lists/*
 
+# Compile and install GCC
+ARG GCC_VERSION=14.2.0
+ARG GCC_COMPILE_ARGS=""
+RUN set -ex; \
+    curl -fL https://ftpmirror.gnu.org/gcc/gcc-${GCC_VERSION}/gcc-${GCC_VERSION}.tar.xz -o /tmp/gcc.tar.xz; \
+    mkdir -p /tmp/gcc; \
+    tar -xf /tmp/gcc.tar.xz -C /tmp/gcc --strip-components=1; \
+    rm /tmp/gcc.tar.xz; \
+    cd /tmp/gcc; \
+    ./contrib/download_prerequisites; \
+    (rm *.tar.* || true); \
+    dpkgArch="$(dpkg --print-architecture)"; GCC_PLATFORM_ARGS=""; \
+    case "${dpkgArch##*-}" in  \
+      armel) GCC_PLATFORM_ARGS="--with-arch=armv4t --with-float=soft";; \
+      armhf) GCC_PLATFORM_ARGS="--with-arch=armv7-a --with-float=hard --with-fpu=vfpv3-d16 --with-mode=thumb";; \
+      i386|i686) GCC_PLATFORM_ARGS="--with-arch=i686";; \
+      amd64) GCC_PLATFORM_ARGS="--enable-multilib --with-abi=m64 --with-arch-32=i686 --with-multilib-list=m32,m64,mx32";; \
+    esac; \
+    gnuArch="$(dpkg-architecture --query DEB_BUILD_GNU_TYPE)"; \
+    ln -s "/usr/include/${gnuArch}/asm" "/usr/include/asm"; \
+    majorVersion="${GCC_VERSION%%.*}"; \
+    ./configure \
```

Comment:
I am against of building gcc everytime. It's heavy process, it make things that supposed to be close to instant to be "1 hour long".

If you really need custom gcc build - let's build it, put somewhere as a binary, and only download it and verify hashes. It's not necessary should be a build that is done by gcc's team, but it can be 3rd party repo

Reply:
gcc is *default* compiler in my ubuntu 24.10, I don't see a reason to build it from scratch. It's not a bleeding-edge-night-build which we need.

Reply:
maybe their's? https://github.com/xpack-dev-tools/gcc-xpack

Reply:
> It's heavy process, it make things that supposed to be close to instant to be "1 hour long".

It doesn't take an hour, it takes ~25 minutes ([build](https://github.com/kwvg/dash/actions/runs/11948255950/job/33305538060)) and efforts have been made to make it cacheable as possible (as mentioned in the PR description, see below, along with all the alternatives explored, see above). Builds without the GCC compile take ~5 minutes ([build](https://github.com/kwvg/dash/actions/runs/11898510091/job/33155166212)), so the slowdown impact in a worst-case scenario (cache-miss) is 20 minutes and the subsequent caching should should give us pull times comparable to regular cache hits.

> * Reducing the packages installed in first `RUN` layer to the minimum needed for GCC and LLVM installations and placing GCC build and install to the second `RUN` layer.

I'm not against building them separately and downloading binaries for them. What about baking in the first two steps into their own Docker container and `FROM`'ing it? It should bypass cache misses and the rebuilds that it would attract.

---

> gcc is default compiler in my ubuntu 24.10, I don't see a reason to build it from scratch. It's not a bleeding-edge-night-build which we need.

This has been addressed in the PR description (see below)

> * Moving the development image base to noble as GCC 14 is available for it ([source](https://packages.ubuntu.com/noble/g++-14)) (but means we aren't testing against the lowest supported LTS).

---

> maybe their's? https://github.com/xpack-dev-tools/gcc-xpack

I had suggested this to Pasta and he seemed lukewarm towards the idea (@PastaPastaPasta, thoughts?)